### PR TITLE
Drum push + metadata file

### PR DESCRIPTION
**Pull request by 4dahalibut**
# This repository is public. Do not put here any private DataRobot or customer's data: code, datasets, model artifacts, .etc.

## Summary

Lets people add a training or inference model from the CLI. These models are configured with a yaml file called model-metadata.yaml. 
## Rationale
This will make it easier for users to iterate on new models by not necessitating going through the UI to add a new model. It will also allow us to sneak in some validation if we centralize users to using this tool to push models into DataRobot.

### CODE REVIEW

File: model_templates/training/python3_sklearn/push.yaml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+model:
+    type: training
+    targetType: binary
+    environmentID: 5e8c889607389fe0f466c72d
+client:
+    endpoint: https://staging.datarobot.com/api/v2/
+    token: NWVhNzNlMDQ3OTM1OWUxODQ3ZWMxMWFiOk5EdGxPN1lEYWU0NUtFc3lZMjhmMzhHTFdPWGI5L2ovT1dRajA5Yk5XWGs9
```
Comment (meltzg) [inline model_templates/training/python3_sklearn/push.yaml:7]: is this your token?

Comment (4dahalibut) [reply model_templates/training/python3_sklearn/push.yaml]: of course not.

**Comment by 4dahalibut** [comment]
> why yaml? I think JSON might be more universally understood among non-engineers

Yaml is less ugly.

File: model_templates/training/python3_sklearn/push.yaml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+model:
+    name: drumpush-regression
+    type: training
+    targetType: regression
+    environmentID: 5e8c889607389fe0f466c72d
+client:
+    endpoint: https://staging.datarobot.com/api/v2/
+    token: NWVhNzNlMDQ3OTM1OWUxODQ3ZWMxMWFiOk5EdGxPN1lEYWU0NUtFc3lZMjhmMzhHTFdPWGI5L2ovT1dRajA5Yk5XWGs9
```
Comment (liororama) [inline model_templates/training/python3_sklearn/push.yaml]: Hi, 
Wait a sec.. lets have a discussion about metdata in DRUM models before we add anything.. 
This addition requires a PBMP or at least a design doc.. 
Please lets not do this  in  a PR without  a design discussion.
@yakov-g

File: model_templates/training/python3_sklearn/push.yaml [hunk 3]
```diff
@@ -0,0 +1,8 @@
+model:
+    name: drumpush-regression
+    type: training
+    targetType: regression
+    environmentID: 5e8c889607389fe0f466c72d
+client:
```
Comment (liororama) [inline model_templates/training/python3_sklearn/push.yaml:6]: Client part should not be part of the metadata since this is not related to the model  itself.
Probably the endpoint and  token should be  provided via separate arguments to the drum or env

drum push -d /opt/models/my_model_1 --endpoint XXXX --token YYYY

File: model_templates/training/python3_sklearn/push.yaml [hunk 4]
```diff
@@ -0,0 +1,8 @@
+model:
+    name: drumpush-regression
+    type: training
+    targetType: regression
+    environmentID: 5e8c889607389fe0f466c72d
```
Comment (liororama) [inline model_templates/training/python3_sklearn/push.yaml:5]: Using an ID is not really user facing.. do we have other options? Like using a name?

File: model_templates/training/python3_sklearn/push.yaml (hunk 3 above)
Comment (yakov-g) [reply model_templates/training/python3_sklearn/push.yaml:6]: and env ID also should go just to the CLI param

Comment (yakov-g) [reply model_templates/training/python3_sklearn/push.yaml:6]: drum push will create a model?
What will happen if I call `drum push` again? model will be existing model will be updated or another model will be  created?
Maybe we need to add model Id? And probably it should be just a string?

File: model_templates/training/python3_sklearn/push.yaml (hunk 2 above)
Comment (yakov-g) [reply model_templates/training/python3_sklearn/push.yaml]: Although we discussed some aspects with @4dahalibut , but looks like @liororama is right and it is just too big for a PR.

Comment (4dahalibut) [reply model_templates/training/python3_sklearn/push.yaml]: I wanted this PR to be the design discussion

Comment (4dahalibut) [reply model_templates/training/python3_sklearn/push.yaml]: I could schedule a meeting if you'd like

File: model_templates/training/python3_sklearn/push.yaml (hunk 4 above)
Comment (4dahalibut) [reply model_templates/training/python3_sklearn/push.yaml:5]: We can use a name but then people have to spell it right which is risky

Comment (4dahalibut) [reply model_templates/training/python3_sklearn/push.yaml:5]: We could probably do some interesting regex to accomplish this though

Comment (klichukb) [reply model_templates/training/python3_sklearn/push.yaml:5]: Often in such cases slugs are used as user-facing identifiers.
This repo already has unique slugs internally ("python3_keras", "python3_sklearn"), which could be easy to refer to.
And then if you look, push.yaml already lives under "python3_sklearn" env identifier, so its more or less aware.
Downside: I havent seen similar patterns in DR except maybe layout/overlay names in DRCA, + maintaining constraints within DR.

Comment (klichukb) [reply model_templates/training/python3_sklearn/push.yaml:5]: (Remembered that we slugify organization names but use them afaik internally)

File: model_templates/training/python3_sklearn/push.yaml (hunk 3 above)
Comment (klichukb) [reply model_templates/training/python3_sklearn/push.yaml:7]: > > why yaml? I think JSON might be more universally understood among non-engineers
> 
> Yaml is less ugly.

agree that yaml is more user friendly and readable

File: model_templates/inference/python3_sklearn/model-metadata.yaml [hunk 5]
```diff
@@ -0,0 +1,10 @@
+name: drumpush-regression
+type: inference
+targetType: regression
+modelID: 5f1f15a4d6111f01cb7f91fd
+environmentID: 5e8c889607389fe0f466c72d
+inferenceModel:
+  targetName: MEDV
+validation:
+  inputData: hi
```
Comment (yakov-g) [inline model_templates/inference/python3_sklearn/model-metadata.yaml:9]: Is this expected to be a path?

Comment (yakov-g) [reply model_templates/inference/python3_sklearn/model-metadata.yaml:9]: Could you add an example with class labels? Maybe even in regression, but null values.

File: tests/drum/run-drum-tests-in-container.sh [hunk 6]
```diff
@@ -38,7 +38,7 @@ source $GIT_ROOT/tests/drum/integration-helpers.sh
 
 cd $GIT_ROOT || exit 1
 
-CMRUNNER_REQUIREMENT_PATH=$GIT_ROOT/custom_model_runner/requirements.txt
+CMRUNNER_REQUIREMENT_PATH=$GIT_ROOT/custom_model_runner/requirements_dev.txt
```
Comment (yakov-g) [inline tests/drum/run-drum-tests-in-container.sh:41]: Why?

File: tests/drum/run-drum-tests-in-container.sh [hunk 7]
```diff
@@ -47,9 +47,9 @@ build_docker_image_with_cmrun tests/fixtures/cmrun_docker_env \
                                $CMRUNNER_REQUIREMENT_PATH || exit 1
 
 echo
-echo "Installing the requirements for all tests:  $GIT_ROOT/requirements.txt"
+echo "Installing the requirements for all tests:  $GIT_ROOT/requirements_dev.txt"
```
Comment (yakov-g) [inline tests/drum/run-drum-tests-in-container.sh:50]: Why?

File: tests/drum/run-drum-tests-in-container.sh [hunk 8]
```diff
@@ -47,9 +47,9 @@ build_docker_image_with_cmrun tests/fixtures/cmrun_docker_env \
...
 cd $GIT_ROOT || exit 1
-pip install -r $GIT_ROOT/requirements.txt
+pip install -r $GIT_ROOT/requirements_dev.txt
```
Comment (yakov-g) [inline tests/drum/run-drum-tests-in-container.sh:52]: Why?

File: tests/drum/locally-run-drum-tests.sh [hunk 9]
```diff
@@ -41,7 +41,7 @@ function check_env_or_create() {
         exit 1
     fi
     . $ENV_ACTIVATE
-    pip install pytest -r ${GIT_ROOT}/requirements.txt
+    pip install pytest -r ${GIT_ROOT}/requirements_dev.txt
```
Comment (yakov-g) [inline tests/drum/locally-run-drum-tests.sh:44]: Why?

File: tests/drum/run-drum-tests-in-container.sh (hunk 6 above)
Comment (4dahalibut) [reply tests/drum/run-drum-tests-in-container.sh:41]: I added a dependency into requirements_test that I needed. I could change these to requirements_test as well...

File: model_templates/inference/python3_sklearn/model-metadata.yaml (hunk 5 above)
Comment (4dahalibut) [reply model_templates/inference/python3_sklearn/model-metadata.yaml:9]: I'm going to follow this up with work to get the validation working.

**Comment by datarobot-jarvis** [comment]
:dollar: Total cost of executing tests in Jenkins: $**8.58** USD.

[What is this all about?](http://bit.ly/2BahK4V)
